### ISSUE TITLE
Cannot tell whether comment has been made for study components because must click on study component to see comment icon

### ISSUE THREAD

**Issue by eplebel** [ScienceCommons/api#155]
As can be seen in the screenshot below (see https://www.curatescience.org/beta/#/articles/60045), currently a user cannot tell whether a comment has been made for any of the study components (i.e., data, materials, pre-registration, disclosure) because you must click on the study component to be able to see the comment icon. Hence, users may easily miss these comments, which is highly suboptimal in terms of user engagement.
  
![components-comment-bubble-hidden](https://cloud.githubusercontent.com/assets/4512335/7107440/a7ecd19e-e133-11e4-8c07-5f9533dd25cb.png)

Any ideas Lee on how to improve this?

**Comment by smack416** [comment]
Hi, Etienne. Sorry for the delay in responding. I had started typing something on Monday but I seemed to have lost it.

Here's our logic:
1. The icons identify whether there is *any* content within each of the components, whether this be docs, details, or comments. A user knows there's content based on the solid, coloured shape. 
2. If there is content, they user can click to see all types of content, including comments (identified by the comment bubble). Therefore, a user knows whether there are comments against the existing content with their first click. 

This was our intent, and we're happy with this discovery process as it aligns to the intent of the user understanding specific content is available when you see a solid, coloured icon. i.e., if you don't know what content is available in the first place, it's irrelevant if there's a comment associated with that unknown content.

From our point of view, the primary consideration from your question is whether or not this content has changed from the last time the user looked. This applies to both new content and new comments. i.e., the example above has two solid, blue icons. If additional content is added to one of those, including a comment, or one of the other two study comments gain content and turn solid as a result, I won't know as a returning visitor to this study. In our view, the best way to address this would be to allow users to "Follow" the study and have updates pushed to their notification stream when changes are made, with links to jump straight to that new content. For now, the change history of the study should cover this.

**Comment by eplebel** [comment]
>Therefore, a user knows whether there are comments against the existing content with their first click.
Exactly, but they need to click to see whether there's a comment in relation to <i>any</i> of the available content. This is suboptimal because there will often be many studies, meaning a user would have to click many many times to see if there are any comments (e.g., more than 10 times for an article with 5 studies x 2 components/study, which will be fairly typical) . For example, see https://www.curatescience.org/beta/#/articles/49900

A typical user will be interested in knowing there's at least 1 comment (and the number of comments) for any of the available content for any of the components.

**Comment by smack416** [comment]
Let's consider the following:
If there's no context to the comment, then it serves no purpose. Therefore, you need to relate the comment to a specific piece of content. In this case, you first need to know what content is available before you would care whether someone commented on that content. The comment by itself is meaningless. 

So you select the study component icon to see the content therein and you then know if there's a comment related to that piece of content. If that piece of content is important to you, then you'll likely want to see what the comment is all about, but you can't make that distinction until you know what piece of content the comment relates to.

**Comment by eplebel** [comment]
>If there's no context to the comment, then it serves no purpose.
This is simply not true. Imagine you're a user (i.e., a researcher) who is actually using the website and is browsing the Bargh & Shalev (2012) article (https://www.curatescience.org/beta/#/articles/49900); this is a real scientific article, with real replications, real data, and a real comment! You notice immediately that all the Donnellan et al. (2014) replication studies have available data and materials (this is a constant across these studies). In this (typical) case, you're equally interested in the data/materials for any of these replication studies. If, however, you knew that there was a comment for one of these studies in particular, you'd click on it to look into what the issue being discussed is about!

**Comment by smack416** [comment]
I wouldn't care about the comment unless I cared about the content, at which point I'd see the comment because I'd be looking at the list of content and see the comment bubble. 

If you and others are interested in seeing all discussion around a paper at all times, whether or not it's related to the content currently on screen, then you need to turn the general comment thread below into a stream of user activity which you can click to reveal within its context or have an activity stream that users can view/subscribe to. 

In general, this is a power user feature to see all comments all the time, and you're talking about bubbling information to the surface which is out of context and therefore will be a distraction to most.

**Comment by eplebel** [comment]
OK but I just gave you an actual use case where your logic doesn't work for an actual researcher user (not a hypothetical user using abstract idealized workflow). 

Here's another use case: https://www.curatescience.org/beta/#/articles/59309 , for which your logic also doesn't work.

As a researcher interested in racial bias, I come to this article page and notice that both replication studies each have data, materials, and pre-registration components. I care about ALL of the content in these studies EQUALLY, hence, seeing a comment bubble for one of the components specifically (or even just seeing a comment bubble for one of the specific studies) will lead me dig in and learn more about the replication studies and engage with the website. If users have to click to see comment bubbles, they will often miss interesting comments and become frustrated.

Admittedly, I likely didn't sufficiently elaborate on such use cases during the design stage, because this stuff is complex and difficult to reason through in the abstract. And we didn't really have the ability to build-revise-and-build-again because our development was too slow given that we only had volunteer coders helping us at the time.

One potential solution is to display a comment bubble at the study-level (to the right of the authors and study number), which would indicate if there are ANY comments for ANY of the components/elements of a study. Clicking on the comment bubble would then bring the user to the comment in context of the component/element of the study. This approach would also work for the row-based view (http://etiennelebel.com/rowprototype/index.html#articlepage_page), which will eventually be implemented during the Round 2 enhancements stage.

**Comment by smack416** [comment]
I don't have any access to the article links you've been sharing (blogged by an invite page which just signs me up to Mailchimp), so perhaps I'm missing something because I'm only seeing what you're writing, but I'm still not following how there's a disconnect here. If you're interested in all the content, then you would click to see the content at which point you would see that there's a related comment. Knowing that there's a comment before you know what the content is wouldn't be an incentive to click because the comment is related specifically to the content you have yet to see.

But, if you want to support general commentary at the individual study level, this is a level of comments that hasn't been considered.

**Comment by eplebel** [comment]
i just invited you via your @sayyeah.com email address, though you'll need a google account to log in I believe

**Comment by smack416** [comment]
No luck. I’m just getting kicked back to the signup page when I log in with Google.


> On Apr 16, 2015, at 2:45 PM, Etienne P. LeBel <notifications@github.com> wrote:
> 
> i just invited you via your @sayyeah.com email address, though you'll need a google account to log in I believe
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/ScienceCommons/api/issues/155#issuecomment-93812072>.
> 


Cheers!
Lee

647.898.9694

i need sugar; music and art
http://ineedsugar.com <http://ineedsugar.com/>

nathan lee; mixes and musical treats
http://nathanlee.ca <http://nathanlee.ca/>

feed lee; random musings
http://leedale.ca <http://leedale.ca/>

**Comment by eplebel** [comment]
Oh right, that's because of open issue #148.

@alexkyllo, can you take a quick look and see if you can figure out the invite problem (#148)?

**Comment by alexkyllo** [comment]
Yes I will try and troubleshoot this. What's your e-mail address smack416? I need to see if it's in the database or not.

**Comment by smack416** [comment]
Ah...lee@sayyeah.com, though Google may be implying it's lee@yousayyeah.com

**Comment by alexkyllo** [comment]
Yeah, that is what's going on. In the database there is an Invite with one of your e-mail addresses but no User with either of your e-mail addresses. The code that creates Users, checks that your OAuth e-mail (your google+ or gmail address) matches the e-mail address your invite was sent to, and errors before creating the User if the two addresses don't match.

**Comment by eplebel** [comment]
Cool, thanks for looking into this and figuring out the problem!
Can you go ahead and fix the issue (#148) by removing that equality constraint? I don't see how that's a necessary nor valuable feature (though I could be missing something...).
In the meantime Lee, can you send me your gmail address and then I can invite you with that email address?

**Comment by alexkyllo** [comment]
Well, the e-mail address matching is how it tells whether a user has been invited or not, so if I just remove that equality test it will have no way to know that the person you invited and the person signing up are the same person. So it will just allow anyone to create an account regardless of whether you've invited them or not. Is that what you want?

**Comment by eplebel** [comment]
Oh OK, no please leave as is (sorry I misunderstood!).

**Comment by alexkyllo** [comment]
OK. So the workaround for now is to inform users they must log with Google using the same e-mail address you sent the invite to. 

A good enhancement would be to add invite tokens so that the URL they get in the invite e-mail is unique to them and expires after a period of time. That way the app would know if the user signing up has been invited or not, regardless of how they authenticate with Google, and you'd be able to remove that email equality test.

But that's a nontrivial feature, and by that time you might be at the point where you want to remove the invite system altogether and open it up to everyone.

**Comment by smack416** [comment]
Thanks.

My gmail address then is probably registering as lee@yousayyeah.com <mailto:lee@yousayyeah.com> as lee@sayyeah.com <mailto:lee@sayyeah.com> is just an alias (we got our primary domain only after years of haggling with a chap in the Netherlands, so it’s just an alias).

> On Apr 19, 2015, at 3:30 PM, Etienne P. LeBel <notifications@github.com> wrote:
> 
> Cool, thanks for looking into this and figuring out the problem!
> Can you go ahead and fix the issue (#148 <https://github.com/ScienceCommons/api/issues/148>) by removing that equality constraint? I don't see how that's a necessary nor valuable feature (though I could be missing something...).
> In the meantime Lee, can you send me your gmail address and then I can invite you with that email address?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/ScienceCommons/api/issues/155#issuecomment-94307360>.
> 


Cheers!
Lee

647.898.9694

i need sugar; music and art
http://ineedsugar.com <http://ineedsugar.com/>

nathan lee; mixes and musical treats
http://nathanlee.ca <http://nathanlee.ca/>

feed lee; random musings
http://leedale.ca <http://leedale.ca/>
